shell/common/shell.cc

```diff
@@ -876,6 +877,77 @@ void Shell::UpdateIsolateDescription(const std::string isolate_name,
   vm_->GetServiceProtocol()->SetHandlerDescription(this, description);
 }
 
+void Shell::SetNeedsReportTimings(bool value) {
+  needs_report_timings_ = value;
+}
+
+void Shell::ReportTimings() {
+  FML_DCHECK(is_setup_);
+  FML_DCHECK(task_runners_.GetGPUTaskRunner()->RunsTasksOnCurrentThread());
+
+  // The engine might have been destructed in ~Shell while the rasterizer is
+  // still alive and calling this from OnFrameRasterized.
+  if (!engine_) {
+    return;
+  }
+
+  auto timings = std::move(unreported_timings_);
+  unreported_timings_ = {};
+  task_runners_.GetUITaskRunner()->PostTask([timings, engine = GetEngine()] {
+    engine->ReportTimings(std::move(timings));
+  });
+}
+
+size_t Shell::UnreportedFramesCount() const {
+  // Check that this is running on the GPU thread to avoid race conditions.
+  FML_DCHECK(task_runners_.GetGPUTaskRunner()->RunsTasksOnCurrentThread());
+  FML_DCHECK(unreported_timings_.size() % FrameTiming::kCount == 0);
+  return unreported_timings_.size() / FrameTiming::kCount;
+}
+
+void Shell::OnFrameRasterized(const FrameTiming& timing) {
+  FML_DCHECK(is_setup_);
+  FML_DCHECK(task_runners_.GetGPUTaskRunner()->RunsTasksOnCurrentThread());
+
+  // The C++ callback defined in settings.h and set by Flutter runner. This is
+  // independent of the timings report to the Dart side.
+  if (settings_.frame_rasterized_callback) {
+    settings_.frame_rasterized_callback(timing);
+  }
+
+  if (!needs_report_timings_) {
+    return;
+  }
+
+  for (auto phase : FrameTiming::kPhases) {
+    unreported_timings_.push_back(
+        timing.Get(phase).ToEpochDelta().ToMicroseconds());
+  }
+
+  // In tests using iPhone 6S with profile mode, sending a batch of 1 frame or a
+  // batch of 10 frames have roughly the same cost of 0.1ms. Sending a batch of
+  // 100 frames costs about 1ms. So 10 seems to be a good choice.
+  if (UnreportedFramesCount() >= 10) {
```

Comment:
might be worth mentioning in the docs for the callback in dart that it's not free (1ms is a lot!)

Reply:
Will add it. BTW, it's 0.1ms instead of 1ms as we're batching for 10 frames instead of 100 frames.

Reply:
I just meant that 100 frames taking 1ms seems like a lot. (I don't doubt the results, just surprised that this is so expensive. I guess it's good we're not enabling this by default!)

Reply:
Ah, yes, 100 frames for 1ms surprised me as well. The tracing shows that converting C++ int64_t to Dart int is costly. 400 conversions (100 frames) could take as much as 1ms. I guess that it would be much faster to directly send bytes instead of List<int> to Dart. But using raw bytes doesn't seem to be a very friendly API...

Reply:
DBC

It looks like tonic::ToDart is very inefficient for std::vector<int64_t>: it creates an ordinary Dart List and converts each int64_t value individually, creating Dart_Handle for each value. Instead, we should consider using typed data (Dart_NewTypedData/Dart_TypedDataAcquireData or maybe even Dart_NewExternalTypedData) to pass large number of int64_t values to Dart.

Reply:
Thank you @alexmarkov ! `Dart_NewTypedData/Dart_TypedDataAcquireData` is much faster as it's essentially just a `memcpy`. I can now do 500 frames in about 0.2ms. Doing 100 frames is well below 0.1ms. I guess I'll change the strategy to be either 1 second or 100 frames. The 1 second threshold would usually trigger before 100 frames unless we're rendering at 120fps...

Can you please help review the last commit of the `Dart_NewTypedData/Dart_TypedDataAcquireData` change?